R/textstat_simil.R

```diff
@@ -13,10 +13,11 @@
 #'   \code{"features"} for word/term features.
 #' @param method method the similarity or distance measure to be used; see
 #'   Details.
-#' @param upper whether the upper triangle of the symmetric \eqn{V \times V}
```

Comment:
`upper` can be renamed `lower` as the new object only have upper triangle by default. I am fine with changing `fcm` in this way.

Reply:
I think it makes more sense to use the argument as we currently use `tri` in `fcm()`, which is to restrict the output to the upper triangle only. So `upper = TRUE` means we return just the upper triangle, otherwise we return all A-B, B-A pairs and in their matrix form, the transpose of the upper triangle, in other words the full symmetric matrix.

So for `fcm()`, we change `tri` to `upper`, which returns just the upper triangle, which is the same behaviour as `tri`. (We trap `tri` and print a deprecation warning, but change the argument name in the function.)

For this branch of `textstat_simil()`, we restore the `upper` argument but change its meaning from the previous `dist`-like functionality that appeared to return just the lower triangle, to the behaviour similar to `fcm()`.

Reply:
I added back the the `upper = TRUE` as a default. It works like this:

``` r
library("quanteda")
mt <- dfm(corpus_subset(data_corpus_inaugural, Year > 2006)) %>%
  dfm_trim(min_termfreq = 10)

tstat <- textstat_simil(mt, upper = TRUE, diag = TRUE)
tstat
##    document1  document2 similarity
## 1 2013-Obama 2009-Obama  0.9681414
## 2 2017-Trump 2009-Obama  0.9295036
## 3 2017-Trump 2013-Obama  0.9152045
## 4 2009-Obama 2009-Obama  1.0000000
## 5 2009-Obama 2013-Obama  0.9681414
## 6 2013-Obama 2013-Obama  1.0000000
## 7 2009-Obama 2017-Trump  0.9295036
## 8 2013-Obama 2017-Trump  0.9152045
## 9 2017-Trump 2017-Trump  1.0000000
as.matrix(tstat)
##            2009-Obama 2013-Obama 2017-Trump
## 2009-Obama  1.0000000  0.9681414  0.9295036
## 2013-Obama  0.9681414  1.0000000  0.9152045
## 2017-Trump  0.9295036  0.9152045  1.0000000
as.list(tstat)
## $`2009-Obama`
## 2009-Obama 2013-Obama 2017-Trump 
##  1.0000000  0.9681414  0.9295036 
## 
## $`2013-Obama`
## 2013-Obama 2009-Obama 2017-Trump 
##  1.0000000  0.9681414  0.9152045 
## 
## $`2017-Trump`
## 2017-Trump 2009-Obama 2013-Obama 
##  1.0000000  0.9295036  0.9152045

tstat <- textstat_simil(mt, upper = TRUE, diag = FALSE)
tstat
##    document1  document2 similarity
## 1 2013-Obama 2009-Obama  0.9681414
## 2 2017-Trump 2009-Obama  0.9295036
## 3 2017-Trump 2013-Obama  0.9152045
## 5 2009-Obama 2013-Obama  0.9681414
## 7 2009-Obama 2017-Trump  0.9295036
## 8 2013-Obama 2017-Trump  0.9152045
as.matrix(tstat) # always include the diagonal of simil(A,A) = 1.0
##            2009-Obama 2013-Obama 2017-Trump
## 2009-Obama  1.0000000  0.9681414  0.9295036
## 2013-Obama  0.9681414  1.0000000  0.9152045
## 2017-Trump  0.9295036  0.9152045  1.0000000
as.list(tstat) # (A,A) pairs not included
## $`2009-Obama`
## 2013-Obama 2017-Trump 
##  0.9681414  0.9295036 
## 
## $`2013-Obama`
## 2009-Obama 2017-Trump 
##  0.9681414  0.9152045 
## 
## $`2017-Trump`
## 2009-Obama 2013-Obama 
##  0.9295036  0.9152045

tstat <- textstat_simil(mt, upper = FALSE, diag = TRUE)
tstat
##    document1  document2 similarity
## 1 2009-Obama 2009-Obama  1.0000000
## 2 2009-Obama 2013-Obama  0.9681414
## 3 2013-Obama 2013-Obama  1.0000000
## 4 2009-Obama 2017-Trump  0.9295036
## 5 2013-Obama 2017-Trump  0.9152045
## 6 2017-Trump 2017-Trump  1.0000000
as.matrix(tstat) # upper is NA
##            2009-Obama 2013-Obama 2017-Trump
## 2009-Obama  1.0000000         NA         NA
## 2013-Obama  0.9681414  1.0000000         NA
## 2017-Trump  0.9295036  0.9152045          1
as.list(tstat) # same as with upper = TRUE
## $`2009-Obama`
## 2009-Obama 2013-Obama 2017-Trump 
##  1.0000000  0.9681414  0.9295036 
## 
## $`2013-Obama`
## 2013-Obama 2009-Obama 2017-Trump 
##  1.0000000  0.9681414  0.9152045 
## 
## $`2017-Trump`
## 2017-Trump 2009-Obama 2013-Obama 
##  1.0000000  0.9295036  0.9152045
as.dist(as.matrix(tstat)) # works
##            2009-Obama 2013-Obama
## 2013-Obama  0.9681414           
## 2017-Trump  0.9295036  0.9152045

tstat <- textstat_simil(mt, upper = FALSE, diag = FALSE)
tstat
##    document1  document2 similarity
## 2 2009-Obama 2013-Obama  0.9681414
## 4 2009-Obama 2017-Trump  0.9295036
## 5 2013-Obama 2017-Trump  0.9152045
as.matrix(tstat) # upper is NA
##            2009-Obama 2013-Obama 2017-Trump
## 2009-Obama  1.0000000         NA         NA
## 2013-Obama  0.9681414  1.0000000         NA
## 2017-Trump  0.9295036  0.9152045          1
as.list(tstat) # same as with upper = TRUE
## $`2009-Obama`
## 2013-Obama 2017-Trump 
##  0.9681414  0.9295036 
## 
## $`2013-Obama`
## 2009-Obama 2017-Trump 
##  0.9681414  0.9152045 
## 
## $`2017-Trump`
## 2009-Obama 2013-Obama 
##  0.9295036  0.9152045
as.dist(as.matrix(tstat)) # works
##            2009-Obama 2013-Obama
## 2013-Obama  0.9681414           
## 2017-Trump  0.9295036  0.9152045
```